File: oldscool/lib/Win.cs
using System;

namespace OldScool 
{
	public enum TitlePos
	{
		TopLeft,
		TopCenter,
		TopRight,
		BottomLeft, 
		BottomCenter,
		BottomRight
	}


	public enum CursorType
	{
		Off,
		Line,
		Block
	}

	// A window part of the screen
	public class Win
	{
		public Win(Vio vio, int x, int y, int w, int h)
		{
			_vio = vio;
			PosX = x;
			PosY = y;
			Width = w;
			Height = h;

			CursorType = CursorType.Off;
			_titleAttr = AttribRole.WinTitle;
			_frame = CreateFrame(FrameType.Single);
			_frameAttr = AttribRole.WinFrame;

			_visible = false;
			_content = new int [w, h, 2];
			_background = new int [w, h, 2];
			_textAttr = AttribRole.WinText;
		

			Clear();
		}

		public void SetBackground(AttribRole attribRole)
		{
			
		}
		
		
		public int PosX { get; } 
		public int PosY { get; } 
		public int Width { get; }
		public int Height { get; }

		int CenterCol(int len=0)
		{
			return (Width-len) / 2;
		}
		int CenterRow(int height=0)
		{
			return (Height-height) / 2;
		}








		Vio _vio;
		int _curX; // X-Position des Cursors
		int _curY; // Y-Position des Cursors
		int [,,] _content;
		int [,,] _background;
		AttribRole _textAttr;
//  INT           iTitlePos;                       /* Position des Titels     */
//  WORD          wFlags;                          /* Die Fenster-Flags       */
//  WORD          wAttrShadow;                     /* Attribut des Fensterschattens */





		public CursorType CursorType { get; set; }

		#region Title

		string _title;
		TitlePos _titlePosition;
		AttribRole _titleAttr;
		public string Title { get => _title; }
		public TitlePos TitlePosition { get => _titlePosition; }
		public void SetTitle(string title, TitlePos position, AttribRole attribute=AttribRole.WinTitle) { 
			_title = title;
			_titlePosition = position; 
			_titleAttr = attribute;
			DrawFrame(); 
			DrawTitle(); 
		}

		#endregion Title

		#region Frame

		IFrame _frame;
		AttribRole _frameAttr;

		public FrameType FrameType { get => _frame!=null ? _frame.Type : FrameType.None; }
		public void SetFrame(FrameType frameType, AttribRole attribute=AttribRole.WinFrame)
		{
			_frame = CreateFrame(frameType);
			_frameAttr = attribute;
			DrawFrame();
			DrawTitle();
		}

		#endregion Frame

    bool _visible;
    public bool Visible {
      get => _visible;
      set {
        if( _visible==value ) return; 
        _visible = value;
        if( _visible ) { 
          SaveWin();
          DrawFrame(); 
          DrawTitle(); 
          Redraw();
        } else {
          RestoreWin();
        }
      }
    }

	public void Remove() {}


    public void Clear()
    {
      var line = new string(' ', Width-2);
      for(var i=0; i<Height-2; i++) {
        Ssa(0, i, line, _textAttr);  
      }
    }

    public void Sz(int col, int row, char ch)
    {
	    if( Visible )
		    _vio.Sz(PosX+col+1, PosY+row+1, ch);
	    _content[col,row,0] = ch;
    }

    public void Sza(int col, int row, char ch, AttribRole attr)
    {
      if( Visible )
        _vio.Sza(PosX+col+1, PosY+row+1, ch, attr);
      _content[col,row,0] = ch;
      _content[col,row,1] = (int) attr;
    }

    public void Ss(int col, int row, string text)
    {
	    if( Visible )
		    _vio.Ss(PosX+col+1, PosY+row+1, text);
	    for(int i=0; i<text.Length; i++) {
		    _content[col+i,row,0] = text[i];
	    }
    }

    public void Ssa(int col, int row, string text, AttribRole attr)
    {
      if( Visible )
        _vio.Ssa(PosX+col+1, PosY+row+1, text, attr);
      for(int i=0; i<text.Length; i++) {
        _content[col+i,row,0] = text[i];
        _content[col+i,row,1] = (int) attr;
      }
    }

    public void Printf(string text)
    {
      var maxCol = Width-2;
      var maxRow = Height-2;
      var col = _curX;
      var row = _curY;
      for(var i=0; i<text.Length; i++)
      {
        switch( text[i] )
        {
          case '\b':
            if( col>0 ) col--;
            break;
          case '\r':
            col = 0;
            break;
          case '\n':
            if( row==maxRow ) 
              ; //ScrollUp();
            else
              row++;
             col = 0;
            break;
          default:
            Sza( col, row, (char)text[i], _textAttr );
            if( ++col == maxCol ) {
              col = 0;
              if( row==maxRow )
                ; //ScrollUp();
              else
                row++;
            }
            break;
        }
      }
      _curX = col;
      _curY =row;
    }

    void Redraw() 
    {
      for(var i=0; i<Width-2; i++) {
        for(var n=0; n<Height-2; n++) {
          _vio.Sza(PosX+i+1, PosY+n+1, (char)_content[i,n,0], (AttribRole) _content[i,n,1]);
        }
      }
    }



    void RestoreWin()
    {
      for(var i=0; i<Width; i++) {
        for(var n=0; n<Height; n++) {
           _vio.Sza(i+PosX, n+PosY, (char)_background[i, n, 0], (AttribRole) _background[i, n, 1]);
        }
      }
    }

    void SaveWin()
    {
      for(var i=0; i<Width; i++) {
        for(var n=0; n<Height; n++) {
          _background[i, n, 0] = _vio.Lz(i+PosX, n+PosY);
          _background[i, n, 1] = _vio.La(i+PosX, n+PosY);
        }
      }
    }


		void DrawTitle()
		{
			if( !Visible ) return;
			if( string.IsNullOrEmpty(Title) ) return;

			var col = 0;
			var row = PosY;
 
			switch(TitlePosition)
			{
				case TitlePos.BottomLeft:
					row = PosY + Height -1;
					col = PosX + 1;
					break;
				case TitlePos.BottomCenter:
					row = PosY + Height -1;
					col = PosX + CenterCol(Title.Length);
					break;
				case TitlePos.BottomRight:
					row = PosY + Height -1;
					col = PosX + Width - Title.Length - 1;
					break;
				case TitlePos.TopLeft:
					col = PosX + 1;
					break;
				case TitlePos.TopCenter:
					col = PosX + CenterCol(Title.Length);
					break;
				case TitlePos.TopRight:
					col = PosX + Width - Title.Length - 1;
					break;
			}
			_vio.Ssa( col, row, Title, _titleAttr );
		}

		void DrawFrame()
		{
			if( !Visible ) return;
			var width = PosX+Width-1;
			var height = PosY+Height-1;
			_vio.Sza( PosX, PosY, _frame.TopLeft, _frameAttr );
			_vio.Sza( width, PosY, _frame.TopRight, _frameAttr );
			_vio.Sza( PosX, height, _frame.BottomLeft, _frameAttr );
			_vio.Sza( width, height, _frame.BottomRight, _frameAttr );
			for(var i=PosX+1; i<width; i++)
			{
				_vio.Sza( i, PosY, _frame.HLine, _frameAttr );
				_vio.Sza( i, height, _frame.HLine, _frameAttr );
			}
			for(var i=PosY+1; i<height; i++)
			{
				_vio.Sza( PosX, i, _frame.VLine, _frameAttr );
				_vio.Sza( width, i, _frame.VLine, _frameAttr );
			}
		}

//		void DrawShadow()
//		{
//			if( !Visible ) return;
//			var stop = PosY+Height-1;
//			for(var i=0; i<=stop; i++)
//				Vio.Sa( PosX, i, _attrShadow );
/*
  if( wStart < VioGetMaxCol() )    
  {
   wStop  = win->wPosY+win->wHeight; 
   {
    wStop  = win->wPosY+win->wHeight;
    for( i=win->wPosY+1; i<=wStop && i<VioGetMaxRow(); i++ )
    {
     VioSa( wStart, i, win->wAttrShadow );
     if( wStart<VioGetMaxCol()-1 )
      VioSa( wStart+1, i, win->wAttrShadow );
    }
    wStart = win->wPosY+win->wHeight;
    if( wStart < VioGetMaxRow() )
    {
     wStop  = win->wPosX+win->wWidth;
     for( i=win->wPosX+1; i<wStop && i< VioGetMaxCol(); i++ )
      VioSa( i, wStart, win->wAttrShadow );
    }
   }
  }
 */
		IFrame CreateFrame(FrameType type) 
		{
			if( _frame!=null ) {
				if( _frame.Type==type )
					return _frame;
			}
			return Frame.CreateFrame(type);
		}
	}
}

File: oldscool.cpp/oldscool/Win.cpp
#include <list>
#include <algorithm>
#include <ncursesw/ncurses.h>
#include <cassert>
#include "Win.h"
#include "Palette.h"

namespace OldScool {

	std::list<Win*> _windows;

	Win::Win(const Vio& vio, int x, int y, int width, int height)
	: _vio(vio)
	, _window(0)
	, _content(0)
	, _textAttr(AttribRole::WinText)
	, _frameAttr(AttribRole::WinFrame)
	, _frame(nullptr)
	, _titlePos(TitlePos::TopCenter)
	, _titleAttr(AttribRole::WinTitle)
	{
		assert( x>0 );
		assert( y>0 );
		if( x+width>vio.getColumns()-1 ) width = vio.getColumns()-1;
		if( y+height>vio.getRows()-1 ) height = vio.getRows()-1;
		_window = newwin(height+2, width+2, y, x);
		_content = subwin(_window, height, width, y+1, x+1);
		_frame = createFrame(FrameType::Single);
		clear();
	}

	Win::~Win() {
		if( _frame )
			delete _frame;
		if( _content )
			delwin(_content);
		if( _window )
			delwin(_window);
		hide();
	}

	int Win::getX()
	{
		return _content->_begx;
	}

	int Win::getY()
	{
		return _content->_begy;
	}

	int Win::getWidth()
	{
		return _content->_maxx;
	}

	int Win::getHeight()
	{
		return _content->_maxy;
	}

	bool Win::isVisible() {
		bool ret = (std::find(_windows.begin(), _windows.end(), this) != _windows.end());
		// TVio::debug(QString("TWin::isVisible %1: %2").arg(objectName()).arg(ret));
		return ret;
	}

	void Win::setFrame(FrameType frame, int attr) {
		_frame = createFrame(frame);
		_frameAttr = attr;
		showFrame( isVisible() );
		showTitle( isVisible() );
	}

	void Win::setTitle(const string& title, TitlePos pos, int attr) {
		_title = title;
		_titlePos = pos;
		_titleAttr = attr;
		showFrame( isVisible() );
		showTitle( isVisible() );
	}

	void Win::show()
	{
		// TVio::debug(QString("TWin::show, showing %1").arg(objectName()));
		_windows.push_back(this);
		update();
//		showWins();
	}

	void Win::hide()
	{
		// TVio::debug(QString("TWin::hide, hiding %1").arg(objectName()));
		::touchwin(stdscr);
		::refresh();
		_windows.remove(this);
		showWins();
/*
		if( m_Parent && m_Parent->m_Win ) {
//			TVio::debug("TWin::hide, touching parent");
			::touchwin(m_Parent->m_Win);
			::wrefresh(m_Parent->m_Win);
		} else {
//			TVio::debug("TWin::hide, touching screen");
			::touchwin(stdscr);
			::refresh();
		}
		m_Hidden = true;
*/
	}

	void Win::clear()
	{
		//::wattrset(_window, _vio.getPalette().get(_textAttr));
		//::wclear(_window);
		// ::wrefresh(_window);
		::wbkgdset(_content, _vio.getPalette().get(_textAttr));
		//::wattrset(_content, _vio.getPalette().get(_textAttr));
		::wclear(_content);
		//if( isVisible() ) {
		//	::wrefresh(_content);
		//}
		showFrame( isVisible() );
	}

	void Win::print(const string& str)
	{
//		wattrset(m_Sub, m_Win->);
		::waddstr(_content, str.c_str());
        if( isVisible() ) ::wrefresh(_content);
	}

    void Win::sz(int x, int y, char ch)
    {
        mvwaddch(_content, y, x, ch);
        if( isVisible() ) ::wrefresh(_content);
    }

    void Win::sza(int x, int y, char ch, int col)
    {
        // ::wattrset(_content, _vio.getPalette().get(col));
        // wmove(_Sub, y, x);
        ::wattron(_content, col);
        mvwaddch(_content, y, x, ch);
        ::wattroff(_content, col);
        if( isVisible() ) ::wrefresh(_content);
    }

    void Win::ss(int x, int y, const string& str)
    {
        mvwaddstr(_content, y, x, str.c_str());
        if( isVisible() ) ::wrefresh(_content);
    }

    void Win::ssa(int x, int y, const string& str, int col)
    {
        // ::wattrset(_content, _vio.getPalette().get(col));
        // :: wattron(_content, _vio.getPalette().get(col));
        ::wattron(_content, col);
        // wmove(m_Sub, y, x);
        // color_set(_vio.getPalette().get(col), 0);
        //::color_set(col, 0);
        mvwaddstr(_content, y, x, str.c_str());
        ::wattroff(_content, col);
        if( isVisible() ) ::wrefresh(_content);
        // wattroff(m_Sub, m_Vio.getColors().getUser(col));
    }

	void Win::showWins()
	{
		for(auto const& win : _windows)
			win->update();
	}

	Frame* Win::createFrame(FrameType type) {
		if( _frame!=nullptr ) {
			if( _frame->getType()==type )
				return _frame;
			delete _frame;
		}
		return Frame::createFrame(type);
	}

	void Win::showFrame(bool refresh)
	{
		//::wattrset(_window, _vio.getPalette().get(_frameAttr));
		::wattron(_window, _frameAttr);
		/*
		mvwaddch(_window, 0, 0, _frame->topLeft());
		mvwaddch(_window, 0, getWidth()+2, _frame->topRight());
		mvwaddch(_window, getHeight()+2, 0, _frame->bottomLeft());
		mvwaddch(_window, getHeight()+2, getWidth()+2, _frame->bottomRight()); */

		::wborder(_window, _frame->vLine(), _frame->vLine(), _frame->hLine(), _frame->hLine(), _frame->topLeft(), _frame->topRight(), _frame->bottomLeft(), _frame->bottomRight());
        // :: wborder(_window, u'\u2502', u'\u2502', u'\u2500', u'\u2500', u'\u250C', u'\u2510', u'\u2514', u'\u2518');
        // wborder(WINDOW *win, chtype ls, chtype rs, chtype ts, chtype bs, chtype tl, chtype tr, chtype bl, chtype br);
		//::box(_window, 0, 0);
        ::wattroff(_window, _frameAttr);
		// ::wattroff(_window, _vio.getPalette().get(_frameAttr));


		if( _title.empty() )
			showTitle(false);
		if( refresh )
			wrefresh(_window);
	}

	void Win::showTitle(bool refresh)
	{
		int spalte = 2;
		int zeile = _titlePos&1 ? _window->_maxy : 0;
		switch( _titlePos )
		{
			default:
			case TitlePos::TopLeft:
			case TitlePos::BottomLeft:
				break;
			case TitlePos::TopRight:
			case TitlePos::BottomRight:
				spalte = _window->_maxx-_title.length()-1;
				break;
			case TitlePos::TopCenter:
			case TitlePos::BottomCenter:
				spalte = (_window->_maxx-_title.length())/2;
				break;
		}
		// Achtung, sa benutzt m_Sub, nicht m_Win!
		::wmove(_window, zeile, spalte);
		::wattrset(_window, _vio.getPalette().get(_titleAttr));
		::waddstr(_window, _title.c_str());
		//::wattroff(_window, _vio.getPalette().get(_titleAttr));
		if( refresh )
			::wrefresh(_window);
/* PDCurses
  if( boMouse ) MouShow();
*/
	}

	void Win::update()
	{
		// TVio::debug(QString("TWin::update, updating %1").arg(objectName()));
		::touchwin(_window);
		::wrefresh(_window);
		::touchwin(_content);
		::wrefresh(_content);
	}

	int Win::getHotkey(const string& text) {

		auto start = text.find('~');
		if( start==std::string::npos ) return -1;
		auto end = text.find('~', start+1);
		if( end==std::string::npos ) return -1;

		auto buff = text.substr(start+1, end-1);
		std::transform(buff.begin(), buff.end(), buff.begin(), ::toupper);

		if(buff.length()>1) {
			if(buff[0]=='F') {

			}
		}
		INT i;
		for (i = 0; pcText[i] != 0; i++) {
			if (pcText[i] == '~') {
				if (pcText[i + 2] == '~') return (toupper(pcText[i + 1]));
				if (pcText[i + 1] == 'F')
					switch (pcText[i + 2]) {
						case '1':
							switch (pcText[i + 3]) {
								case '0':
									return (T_F10);
								case '1':
									return (T_F11);
								case '2':
									return (T_F12);
							}
							return (T_F1);
						case '2':
							return (T_F2);
						case '3':
							return (T_F3);
						case '4':
							return (T_F4);
						case '5':
							return (T_F5);
						case '6':
							return (T_F6);
						case '7':
							return (T_F7);
						case '8':
							return (T_F8);
						case '9':
							return (T_F9);
					}
			}
		}
		return (-1);
	}

}